### PR TITLE
[Fix](PreparedStatement) nondeterministic functions in prepared statement should not be short circuited

### PR DESCRIPTION
**Pull request by eldenmoon**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Thearas** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -407,8 +407,14 @@ public Expression visitUnboundFunction(UnboundFunction unboundFunction, Expressi
         Pair<? extends Expression, ? extends BoundFunction> buildResult = builder.build(functionName, arguments);
         buildResult.second.checkOrderExprIsValid();
         Optional<SqlCacheContext> sqlCacheContext = Optional.empty();
+        StatementContext statementContext = context.cascadesContext.getStatementContext();
+
+        // If it is prepared statement in EXECUTE phase, record nonDeterministic info
+        if (ConnectContext.get() != null && ConnectContext.get().getCommand() == MysqlCommand.COM_STMT_EXECUTE
+                && !buildResult.second.isDeterministic()) {
+            statementContext.setHasNondeterministicInPreparedStatement(true);
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java]: set this flag for all statement or mv this flag out of statement conctext

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/ExecuteCommand.java [hunk 2]
```diff
@@ -70,14 +70,15 @@ public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
...
+        // If it's not a short circuit query or schema version is different(indicates schema changed) or
+        // has nondeterministic functions in statement, then need to do reanalyze and plan
         boolean isShortCircuit = executor.getContext().getStatementContext().isShortCircuitQuery();
         boolean hasShortCircuitContext = preparedStmtCtx.shortCircuitQueryContext.isPresent();
         boolean schemaVersionMismatch = hasShortCircuitContext
                     && preparedStmtCtx.shortCircuitQueryContext.get().tbl.getBaseSchemaVersion()
                     != preparedStmtCtx.shortCircuitQueryContext.get().schemaVersion;
-        boolean needAnalyze = !isShortCircuit || schemaVersionMismatch || !hasShortCircuitContext;
+        boolean needAnalyze = !isShortCircuit || schemaVersionMismatch || !hasShortCircuitContext
+                        || executor.getContext().getStatementContext().hasNondeterministicInPreparedStatement();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/ExecuteCommand.java:81]: does it work? ExpressionAnalyzer be called in nerieds' analyzer. ExpressionAnalyzer should not be called when come here. so i don't think hasNondeterministicInPreparedStatement will be set correctly.

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/ExecuteCommand.java:81]: i got it. u save the statement context and reuse them between these execute command. it works

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java (hunk 1 above)
Comment (eldenmoon) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java]: done

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -407,6 +407,11 @@ public Expression visitUnboundFunction(UnboundFunction unboundFunction, Expressi
         Pair<? extends Expression, ? extends BoundFunction> buildResult = builder.build(functionName, arguments);
         buildResult.second.checkOrderExprIsValid();
         Optional<SqlCacheContext> sqlCacheContext = Optional.empty();
+
+        if (!buildResult.second.isDeterministic() && context != null && context.cascadesContext != null) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java:411]: context.cascades always not null

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java (hunk 3 above)
Comment (eldenmoon) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java]: done

**Review by morrySnow — APPROVED** [review]

**Review by csun5285 — APPROVED** [review]
LGTM

**Pull request merged**